net/http/inc/THttpLongPollEngine.h

```diff
@@ -15,15 +15,34 @@
 #include "THttpWSEngine.h"
 
 #include <string>
-#include <vector>
+#include <list>
+
+class THttpServer;
 
 class THttpLongPollEngine : public THttpWSEngine {
+friend class THttpServer;
 protected:
-   THttpCallArg *fPoll;              ///!< polling request, which can be used for the next sending
-   std::vector<std::string> fQueue;  ///!< entries submitted to client
+   class QueueItem {
+   public:
+      const void *buf{nullptr}; ///<! raw memory with ownership
+      int len{0};               ///<! size of raw memory
+      std::string msg;          ///<! plain text message
+      QueueItem() = default;
+      QueueItem(const std::string &_msg) : msg(_msg) {}
+      QueueItem(const void *_buf, int _len, const std::string &_msg = "") : buf(_buf), len(_len), msg(_msg) {}
+      ~QueueItem();
+      void reset_buf() { buf = nullptr; }
+   };
+
+   bool fRaw{false};                 ///!< if true, only content can be used for data transfer
+   THttpCallArg *fPoll{nullptr};     ///!< polling request, which can be used for the next sending
+   std::list<QueueItem> fQueue;      ///!< entries submitted to client
```

Comment:
`list` is super inefficient. Please use `vector`. The only reason to use `list` is for frequent mid-insertions and mid-removals. See e.g. https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html

(That's a general comment. `list` should really not be used in our code.)

Reply:
Here is macro which demonstrate why I use list instead of vector

[test.txt](https://github.com/root-project/root/files/1857733/test.txt)

Reason is many reallocation when inserting or deleting elements in vector.
I have implicit memory ownership in the QueueItem class and therefore need to implement copy constructor correctly. Otherwise raw pointer with ownership will be copied wrongly.

To avoid all such complication, I use std::list, which calls constructor once (in emplace_back) and destructor once (when delete first element in queue). 

If I simply replace vector with list, code will fail